sdk/ml/azure-ai-ml/azure/ai/ml/entities/_job/job_service.py

```diff
@@ -94,36 +82,222 @@ def _validate_job_service_type_name(self):
+            status=self.status,
+            port=self.port,
+            properties=updated_properties if updated_properties else self.properties,
+        )
+
     @classmethod
-    def _to_rest_job_services(cls, services: Dict[str, "JobService"]) -> Dict[str, RestJobService]:
+    def _to_rest_job_services(
+        cls,
+        services: Dict[
+            str,
+            Union["JobService", "JupyterLabJobService", "SshJobService", "TensorBoardJobService", "VsCodeJobService"],
+        ],
+    ) -> Dict[str, RestJobService]:
         if services is None:
             return None
 
-        # pylint: disable=protected-access
         return {name: service._to_rest_object() for name, service in services.items()}
 
     @classmethod
-    def _from_rest_object(cls, obj: RestJobService) -> "JobService":
+    def _from_rest_job_service_object(cls, obj: RestJobService):
         return cls(
             endpoint=obj.endpoint,
             job_service_type=JobServiceTypeNames.REST_TO_ENTITY.get(obj.job_service_type, None)
             if obj.job_service_type
             else None,
-            # obj.job_service_type,
-            # nodes="all" if isinstance(obj.nodes, AllNodes) else None,
             nodes="all" if obj.nodes else None,
             status=obj.status,
             port=obj.port,
+            # ssh_public_keys=get_property(obj.properties, "sshPublicKeys"),
             properties=obj.properties,
         )
 
     @classmethod
-    def _from_rest_job_services(cls, services: Dict[str, RestJobService]) -> Dict[str, "JobService"]:
-        """Resolve Dict[str, RestJobService] to Dict[str, JobService]"""
+    def _from_rest_job_services(
+        cls, services: Dict[str, RestJobService]
+    ) -> Dict[
+        str, Union["JobService", "JupyterLabJobService", "SshJobService", "TensorBoardJobService", "VsCodeJobService"]
+    ]:
+        # """Resolve Dict[str, RestJobService] to Dict[str, Specific JobService]"""
         if services is None:
             return None
 
         result = {}
         for name, service in services.items():
-            result[name] = JobService._from_rest_object(service)
+            if service.job_service_type == JobServiceTypeNames.RestNames.JUPYTER_LAB:
+                result[name] = JupyterLabJobService._from_rest_object(service)
+            elif service.job_service_type == JobServiceTypeNames.RestNames.SSH:
+                result[name] = SshJobService._from_rest_object(service)
+            elif service.job_service_type == JobServiceTypeNames.RestNames.TENSOR_BOARD:
+                result[name] = TensorBoardJobService._from_rest_object(service)
+            elif service.job_service_type == JobServiceTypeNames.RestNames.VS_CODE:
+                result[name] = VsCodeJobService._from_rest_object(service)
+            else:
+                result[name] = JobService._from_rest_object(service)
         return result
+
+
+@experimental
+class JobService(JobServiceBase):
```

Comment:
To confirm. This is not a breaking change as if customers have existing code of `job_service = JobService(type="ssh", ...)`, it will still work. Is that right?

Reply:
Yes, this is not a breaking change. JobService(job_service_type="ssh",...) works. A test is available at 
D:\T\Git\azure-sdk-for-python\sdk\ml\azure-ai-ml\tests\dsl\unittests\test_dsl_pipeline_with_specific_nodes.py